src/openforms/submissions/tasks/registration.py

```diff
@@ -1,30 +1,81 @@
 import logging
+import traceback
+from typing import Optional
 
 from django.db import IntegrityError, transaction
 from django.utils.crypto import get_random_string
 
 from openforms.celery import app
+from openforms.logging import logevent
+from openforms.registrations.base import BasePlugin
 from openforms.registrations.tasks import register_submission
 
+from ..constants import RegistrationStatuses
 from ..models import Submission
 
 __all__ = [
     "register_submission",
-    "obtain_submission_reference",
     "set_submission_reference",
+    "pre_registration",
+    "obtain_submission_reference",
 ]
 
+
 logger = logging.getLogger(__name__)
 
 
 @app.task
-@transaction.atomic
-def obtain_submission_reference(submission_id: int) -> str:
-    """
-    Task wrapper for set_submission_reference()
-    """
+def pre_registration(submission_id: int) -> None:
     submission = Submission.objects.get(id=submission_id)
-    set_submission_reference(submission)
+    registration_plugin = get_registration_plugin(submission)
+
+    if not registration_plugin:
+        set_submission_reference(submission)
+        return
+
+    options_serializer = registration_plugin.configuration_options(
+        data=submission.form.registration_backend_options
+    )
+    try:
+        options_serializer.is_valid(raise_exception=True)
+    except Exception as e:
+        submission.save_registration_status(
+            RegistrationStatuses.failed, {"traceback": traceback.format_exc()}
+        )
+        logevent.registration_failure(submission, e, registration_plugin)
+        raise
+
+    registration_plugin.pre_register_submission(
+        submission, options_serializer.validated_data
+    )
+
+
+@app.task
+def obtain_submission_reference(submission_id: int) -> None:
+    submission = Submission.objects.get(id=submission_id)
+    registration_plugin = get_registration_plugin(submission)
+
+    if not registration_plugin:
+        return
+
+    # The registration options were validated in the pre-registration task AND in the registration task, so not
+    # validating again. Possible edge case: the options changed between the registration task and the obtain_reference
+    # task?
+    registration_plugin.obtain_submission_reference(
+        submission, submission.form.registration_backend_options
+    )
```

Comment:
this is going a bit in circles now, can we not leave this original implementation?

`set_submission_reference` wraps around `get_reference_for_submission` which loads the registration plugin for the submission and calls `plugin.get_reference_from_result(result)`, so you now have plugin methods `get_reference_from_result` and `obtain_submission_reference` that just call each other (in a way). Additionally, `obtain_submission_reference` does not return a submission reference but mutates the provided submission, which is unexpected.

Preferably functions/methods have no side-effects and return an output for a given input.

Since `set_submission_reference` is idempotent and takes care of calling the correct plugin, I don't see why this `obtain_submission_reference` task needs changing.